### ISSUE TITLE
how to read output file

### ISSUE THREAD

**Issue by shubham953** [flpmasi/Thermodynamics-Neural-Networks#1]
formate of output weight file

**Comment by filippo-masi** [comment]
Hi shubham953,
I think I don't understand the issue you raised.
Could you be more specific?

**Comment by shubham953** [comment]
I want to know how to read input and output.

**Comment by filippo-masi** [comment]
Input and output data? If so, they are saved in binary, so you can import them as in main.py and print them.

**Comment by shubham953** [comment]
They are takken from abacus or fem?Because this is a great work but when i
run this code i couldnot understand how to get the data.In your youtube
vedio theory was clear but how code is running that was not mentioned can
you suggest me some resources to learn this .

**Comment by filippo-masi** [comment]
They are generated as described in:
Masi, Stefanou, Vannucci, and Maffi-Berthier (2020). "Thermodynamics-based Artificial Neural Networks for constitutive modeling". Journal of the Mechanics and Physics of Solids, 104277.

The input and output data are formatted according to the `call` function in [/classThermodynamicsANNs/ThermodynamicsANNs.py ](https://github.com/flpmasi/Thermodynamics-Neural-Networks/blob/main/classThermodynamicsANNs/ThermodynamicsANNs.py)

Inputs = concatenation of strain, strain increment, stress, and internal variables @ time t
```
un_ε_t_f = tf.slice(un,[0,0],[-1,1])
un_Dε_t_f = tf.slice(un,[0,1],[-1,1])
un_σ_t_f = tf.slice(un,[0,2],[-1,1])
un_α_t_f = tf.slice(un,[0,3],[-1,1])
```
Outputs = [internal variable increment, energy, stress increment, dissipation rate] @ time t+dt

`return nf_Dα_tdt, nf_tdt, nf_Dσ_tdt, nf_d_tdt`

You can also find more in:
Masi, Filippo, and Ioannis Stefanou. "Thermodynamics-based Artificial Neural Networks (TANN) for multiscale modeling of materials with inelastic microstructure." arXiv preprint arXiv:2108.13137 (2021).

**Comment by shubham953** [comment]
Thanks, I'll check it out.

***@***.***

**Comment by shubham953** [comment]
can you share the generated data (FEM) code at gmail
***@***.***

**Comment by filippo-masi** [comment]
In this repository, data used to train TANN have been generated without the use of Finite Element simulations.
The data are related to the material point level and come from the integration of hyper- and hypo-plastic incremental constitutive laws. You can find the latter at https://github.com/flpmasi/h2plasticity

**Comment by shubham953** [comment]
Thank you very much for this support.

**Comment by filippo-masi** [comment]
You're welcome!

We will soon release a repository for the enhanced formalism of TANN, in arXiv:2108.13137, that operates on heterogeneous scales (e.g. FEA)

**Comment by shubham953** [comment]
Please share that repository also as this work is quite interesting to
learn.

**Comment by shubham953** [comment]
From where do I find internal variable data I have gone through your paper but did not get it. Understood how to generate stress,strain, energy data but internal variable are quite confusing. Thanks in advance.

**Comment by shubham953** [comment]
e_t,e_tdt,De_t,ap_t,sg_t = file_inp please tell the respective meaning used in preprocessing
ap_tdt,Dap_t,F_tdt,sg_tdt,Dsg_t,D_tdt = file_out please tell the respective meaning used in preprocessing

**Comment by filippo-masi** [comment]
Hi [shubham953](https://github.com/shubham953),
What do you mean with where do I find internal variables? In this repository, and the corresponding paper, the internal variables are chosen to be the inelastic deformations. 
With reference to your second comment:
`e_t,e_tdt,De_t,ap_t,sg_t `= total strain @ time t, total strain @ time t+dt, inelastic deformation @ time t, stress @ time t
 `ap_tdt,Dap_t,F_tdt,sg_tdt,Dsg_t,D_tdt` = inelastic deformation @ time t+dt, inelastic deformation increment @ time t, free-energy @ time t+dt, stress @ time t+dt, stress increment @ time t, dissipation rate @ time t+dt

**Comment by shubham953** [comment]
we use relative weights for outputs of the network in the loss function. Are they calculated from algorithms or from experience?
Internal variables(Z) or internal coordinates(jeeta) as shown in the paper can be velocity,momentum so as you said inelastic deformation is the internal coordinates or the internal variable.

**Comment by shubham953** [comment]
Can you give the surface overview behind preprocessing or the key points of it like(using get _alpha_B,(1 and 0 for free energy and dissipation rate but std and mean for strain))

**Comment by shubham953** [comment]
Can you suggest methodes to evaluate relative weightage of losses

**Comment by filippo-masi** [comment]
Hi @shubham953,

Sorry for the late reply, here the answer to your questions:


> we use relative weights for outputs of the network in the loss function. Are they calculated from algorithms or from experience? Internal variables(Z) or internal coordinates(jeeta) as shown in the paper can be velocity,momentum so as you said inelastic deformation is the internal coordinates or the internal variable.

A1: Relative weights are used for balancing the learning of several quantities, having the resulting total loss being the weighted sum of the loss of each output. In general, you can use unit relative weights, and the training will work. Nevertheless, what we experienced is that, depending on the amplitude of variation of the outputs, sometimes using appropriate relative weights results in a faster convergence. In that scenario, we have found that the best way for deciding the relative weights is to simply consider the coefficient of variation of each output (the larger the range, the higher the weight is).

A2: The internal coordinates are quantities defined at the microscopic level (microscopic deformations, momentum, velocity, displacements, and so on). The internal variables are an average representation (the output of an encoder, cf. the paper) of those. So, in a sense, you could think of using the microscopic inelastic deformations as internal coordinates and retrieve as internal variables the average inelastic deformations, over the unit cell.


> Can you give the surface overview behind preprocessing or the key points of it like(using get _alpha_B,(1 and 0 for free energy and dissipation rate but std and mean for strain))

A3: This comes usually on how the data are generated. Using the procedure detailed in the paper you referenced, the strain and stress best fit in a normal distribution (so we apply a standardization). On the contrary, all those data that do not fit a normal distribution are usually (in most of the literature) normalized according to a uniform distribution (-1 and 1, or 0 and 1, for positive definite quantities) 

> Can you suggest methodes to evaluate relative weightage of losses

A4: See A1.

**Comment by shubham953** [comment]
Thanks for clearing doubts your precious response helping me to learn.

**Comment by shubham953** [comment]
Why we are doing standard normalization .
will it change our input data ?

**Comment by shubham953** [comment]
i am unable to see encoder-decoder thing in your code. have you used it??

**Comment by shubham953** [comment]
can you more elaborate on this 
Data are generated by identifying an initial, reference state for the material at time t=0 with initial zero strain and stress, and by applying random strain increments with uniform standard distribution, assumed constant over the individual time step, Δt. The randomly generated material states are then divided into two sets of increments, one a time t and the other at time t+Δt. The data sets are randomly shuffled, as the evolution of the material state is fully described by the ISV, and process histories are not needed.
*My confusion is how we can generate time steps for ΔZ=Z+dot(Z)*Δt *
basically how time step implemented

**Comment by shubham953** [comment]
is dot(Z) representing only increment in inelastic deformations (as internal coordinates) in the code while generated?
